lib/engine-universal/src/code_memory.rs

```diff
@@ -71,14 +69,16 @@ impl CodeMemory {
             }) + executable_sections.iter().fold(0, |acc, exec| {
                 round_up(acc + exec.bytes.len(), ARCH_FUNCTION_ALIGNMENT)
             }),
-            page_size,
         ) + data_sections.iter().fold(0, |acc, data| {
             round_up(acc + data.bytes.len(), DATA_SECTION_ALIGNMENT)
         });
 
         // 2. Allocate the pages. Mark them all read-write.
-
-        self.mmap = Mmap::with_at_least(total_len)?;
+        if total_len <= self.mmap.len() {
+            self.unpublish()
+        } else {
+            self.mmap = Mmap::with_at_least(total_len).map_err(|e| e.to_string())?;
```

Comment:
Would be cool to push this error to the caller completely, such that we don't have to worry about *internal* mmap failing.